### ISSUE TITLE
Move "self" up?

### ISSUE THREAD

**Issue by duclet** [qubyte/fetch-ponyfill#24]
The variable "self" is declared on line 10, however, it is being used before that on lines 5-6. My test runner is throwing an error because of this. Can we not move it further up?

**Comment by qubyte** [comment]
It's probably possible to do this by aliasing the global `XMLHttpRequest` and `Promise` properties of self in the outermost closure. You'd need to be careful to ensure that the module still loads in Node.js though, (while you can't use it directly in Node, it might be helpful for testing purposes). If you find you can get the behaviour you need without hurting behaviour (probably) for anyone else then please open a pull request. :)

**Comment by duclet** [comment]
I'm curious as to why it was done like how it is now. I think the only reason it is working right now in the browser is because "self" === "window". Though, looking at the code, I see two "self" here. The first one is the one that === "global", which is the root context. The second is the one that will be modified by the Fetch API polyfill. Can't we just pass in "this" and call it something else so we don't have this issue?

**Comment by qubyte** [comment]
At first the module is using `self` as an alias for the global object to pluck `XMLHttpRequest` and `Promise`. Subsequently, a new variable called `self` is defined in a closure. The whatwg-fetch module is templated into that closure. It also attempts to use `self`, which is now a variable the module controls. This is to prevent the whatwg-fetch code modifying or otherwise having access to the actual global object (allowing it to turn a polyfill into a ponyfill).

`self` is used by both this module and the what-wg module it wraps. In both cases `self` is used as a reliable alias for the global object, both in the browser and in a web worker. It's difficult to guarantee that `this` won't be messed with by module systems.

**Comment by qubyte** [comment]
If you provide some sample code and some background about which runner you're using, perhaps we can find another way to work around the problem you are encountering.

**Comment by duclet** [comment]
I don't know of a modular system that would mess with `this`. Since `self` === `window`, why can't you use `window` instead? I see two reasons for this:
* Since the file is assumed to be for the browser, most people would know what `window` is versus `self` (I'm not sure how many people knows that they are the same).
* It takes away the confusion of there being two `self`s. And since there is no `use strict` in the code, I do wonder if variable hoisting also plays into this in older versions of IE.

Basically what is happening here is I build using WebPack and run the tests via Mocha. When it runs, it complains that `self` (rightly so) does not exists yet on line 5.

**Comment by qubyte** [comment]
`window` does not work in web workers. See: https://developer.mozilla.org/en/docs/Web/API/Window/self

**Comment by duclet** [comment]
Well, that is sad. Then the only way around this I see is to pass it in like how the Fetch API does it. If there is a `self`, pass that in and if not, pass in `this`.
